### ISSUE TITLE
lhs2tex does not work with GHC 7.10.1

### ISSUE THREAD

**Issue by jstolarek** [kosmikus/lhs2tex#40]
lhs2tex does not seem to work with GHC 7.10.1. I tried several scenarios:

1. When I say `cabal install lhs2tex` I get:

```
Downloading lhs2tex-1.18.1...
Configuring lhs2tex-1.18.1...
Failed to install lhs2tex-1.18.1
Build log ( /home/killy/.cabal/logs/lhs2tex-1.18.1.log ):
cabal: Error: some packages failed to install:
lhs2tex-1.18.1 failed during the configure step. The exception was:
user error ('/home/killy/.ghc-sandbox/bin/ghc' exited with an error:

/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:298:26:
Ambiguous occurrence ‘die’
It could refer to either ‘Distribution.Simple.Utils.die’,
imported from ‘Distribution.Simple.Utils’ at
/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:16:1-32
or ‘System.Exit.die’,
imported from ‘System.Exit’ at
/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:26:1-18

/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:334:17:
Ambiguous occurrence ‘die’
It could refer to either ‘Distribution.Simple.Utils.die’,
imported from ‘Distribution.Simple.Utils’ at
/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:16:1-32
or ‘System.Exit.die’,
imported from ‘System.Exit’ at
/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:26:1-18

/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:338:17:
Ambiguous occurrence ‘die’
It could refer to either ‘Distribution.Simple.Utils.die’,
imported from ‘Distribution.Simple.Utils’ at
/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:16:1-32
or ‘System.Exit.die’,
imported from ‘System.Exit’ at
/tmp/lhs2tex-1.18.1-12790/lhs2tex-1.18.1/dist/setup/setup.hs:26:1-18
)
```

2. I cloned the github repo and tried `cabal install`. That fails with:
```
(./ZipStupid.tex [38]) (./ZipMath.tex) [39] [40] (./Guide2.aux) )
(see the transcript file for additional information){/usr/share/texmf/fonts/enc
/dvips/base/8r.enc}</usr/share/texmf/fonts/type1/public/amsfonts/cm/cmitt10.pfb
></usr/share/texmf/fonts/type1/public/amsfonts/cm/cmmi10.pfb></usr/share/texmf/
fonts/type1/public/amsfonts/cm/cmr10.pfb></usr/share/texmf/fonts/type1/public/a
msfonts/cm/cmss10.pfb></usr/share/texmf/fonts/type1/public/amsfonts/cm/cmss9.pf
b></usr/share/texmf/fonts/type1/public/amsfonts/cm/cmssbx10.pfb></usr/share/tex
mf/fonts/type1/public/amsfonts/cm/cmsy10.pfb></usr/share/texmf/fonts/type1/publ
ic/amsfonts/cm/cmsy9.pfb></usr/share/texmf/fonts/type1/public/amsfonts/cm/cmtex
8.pfb></usr/share/texmf/fonts/type1/public/amsfonts/cm/cmtex9.pfb></usr/share/t
exmf/fonts/type1/public/amsfonts/cm/cmtt10.pfb></usr/share/texmf/fonts/type1/pu
blic/amsfonts/cm/cmtt12.pfb></usr/share/texmf/fonts/type1/public/amsfonts/cm/cm
tt8.pfb></usr/share/texmf/fonts/type1/public/amsfonts/cm/cmtt9.pfb></usr/share/
texmf/fonts/type1/public/mathpazo/fplmri.pfb></usr/share/texmf/fonts/type1/publ
ic/amsfonts/symbols/msam10.pfb></usr/share/texmf/fonts/type1/hoekwater/stmaryrd
/stmary9.pfb></usr/share/texmf/fonts/type1/urw/palatino/uplb8a.pfb></usr/share/
texmf/fonts/type1/urw/palatino/uplbi8a.pfb></usr/share/texmf/fonts/type1/urw/pa
latino/uplr8a.pfb></usr/share/texmf/fonts/type1/urw/palatino/uplri8a.pfb>
Output written on Guide2.pdf (40 pages, 324136 bytes).
Transcript written on Guide2.log.
setup: Guide2.log: invalid argument
cabal: Error: some packages failed to install:
lhs2tex-1.18.1 failed during the building phase. The exception was:
ExitFailure 1
```

However `cabal build` by itself works. Running `make install` in a clean tree also works and installs the program.

**Comment by kosmikus** [comment]
Yes, I had made the changes but forgot to release. I'll do that soon. The `cabal install` error you got when building from the repo is mysterious though. I guess I also only tried `cabal build`. Thanks for the report.

**Comment by kosmikus** [comment]
I cannot reproduce the problem with the github repo. A fresh clone from github and `cabal install` seems to work just fine for me. What version of `cabal-install` are you using?

**Comment by jstolarek** [comment]
```
$ cabal --version
cabal-install version 1.22.2.0
using version 1.22.2.0 of the Cabal library
```

**Comment by jstolarek** [comment]
A random guess: perhaps this has something to do with the way file paths or regexps are manipulated by the operating system? Setup.hs seems to do both of these in many places so there's a room for things to go wrong. I'm on 64bit Linux.

**Comment by kosmikus** [comment]
Me too. But I was using a HEAD version of `cabal-install`. Will try again with `1.22.2.0`.

**Comment by kosmikus** [comment]
Sorry, still cannot reproduce. Could you please try to run this again with `cabal install -v3` and paste the relevant part of the log? Thanks for your help.

**Comment by jstolarek** [comment]
Not sure which part is the relevant one, so here's the full log: https://gist.github.com/jstolarek/54e7c042c42494a797b1

**Comment by kosmikus** [comment]
Ok, I think I've tracked this down. I'm actually uncertain now why at first I couldn't reproduce it. It's a subtle combination of things though.

For future reference:

   * `InteractiveGhci.lhs` is part of the documentation and invokes `ghci -fglasgow-exts`.
   * The flag `-fglasgow-exts` is long deprecated, but expands to a number of extensions including `UnicodeSyntax`.
   * In `ghc-7.10`, `ghci` has started using Unicode symbols to print e.g. type signatures when `UnicodeSyntax` is enabled.
   * The output of `ghci` is included in `InteractiveGhci.tex` which in turn is included in the LaTeX file that is the documentation.
   * If the output of `ghci` contains unicode symbols (at this point), then the LaTeX log file will contain error messages that a certain font is lacking a certain character.
   * The character printed in the LaTeX log file is itself a strange unicode symbol though that confuses the decoding functions that are applied in `readFile`.
   * The setup file scans the log file for a "rerun" warning and therefore touches this symbol, causing an IO exception to be triggered.

Fixed by no longer passing `-fglasgow-exts` (or `-XUnicodeSyntax`) to the example in question.

**Comment by jstolarek** [comment]
Wow, that was indeed a subtle chain of events. I'll try later whether this fix works for me.

As to you having problems with reproducing the bug I would be guessing that it might have been caused either by different latex installations on our systems or perhaps different locale.

**Comment by jstolarek** [comment]
I confirmed that the problem is gone.

**Comment by kosmikus** [comment]
Ok, I've just released `lhs2tex-1.19` which should build work ok with `ghc-7.10.1`.
